### PR TITLE
fix: run_function not invalidating when attrs change

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

Something I noticed while reading code in #2273. 

<details> <summary>Backward/forward compatibility checks</summary>

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

</details>

## Changelog

- `modal.Image.run_function` now invalidates when `volume`, `cloud`, or `region` is changed. If an `Image` specifies any of these arguments they will rebuild on client upgrade to this version.

### CODE REVIEW

File: modal/functions.py [hunk 1]
```diff
@@ -932,16 +932,20 @@ async def _load(self: _Function, resolver: Resolver, existing_object_id: Optiona
...
-        # hash. We can't use the cloudpickle hash because it's not very stable.
+        # Used to check whether we should rebuild a modal.Image which uses `run_function`.
         gpus: List[GPU_T] = gpu if isinstance(gpu, list) else [gpu]
         obj._build_args = dict(  # See get_build_def
             secrets=repr(secrets),
             gpu_config=repr([parse_gpu_config(_gpu) for _gpu in gpus]),
             mounts=repr(mounts),
             network_file_systems=repr(network_file_systems),
         )
+        # these key are excluded if empty to avoid rebuilds on client upgrade
```
Comment (mwaskom) [inline modal/functions.py:943]: Will this trigger rebuilds if _not_ empty?

Comment (thundergolfer) [reply modal/functions.py:943]: Yes, but that's unavoidable isn't it?

Comment (mwaskom) [reply modal/functions.py:943]: I guess. Worth highlighting in a changelog note at least.

**Review by mwaskom — APPROVED** [review]
I wonder if there's a way to assert write a test asserting that we've done this if the expand the signature further. Given that you're not updating the `build_args` when they're omitted, it seems tricky...

**Pull request merged**